# SQL Project - Denis Haicov

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

### Instructions for completing the task

- Describe the goals of the study.
- Study the tables (print the first rows).
- Make an SQL query for each of the tasks.
- Output the results of each query in the Notebook.
- Describe your conclusions for each of the tasks.

In [1]:
%pip install psycopg2-binary

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
from IPython.display import Image
Image (url = "https://i.postimg.cc/T3YvPCH9/SQL.png", width=800, height=800)

## Study the tables (print the first rows)

In [4]:
query = """SELECT *
FROM books
LIMIT 5;"""
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query = """SELECT *
FROM authors
LIMIT 5;"""
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
query = """SELECT *
FROM ratings
LIMIT 5;"""
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
query = """SELECT *
FROM reviews
LIMIT 5;"""
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [8]:
query = """SELECT *
FROM publishers
LIMIT 5;"""
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

After I have oppened and data I ran all the dataframes to see what I am working with.

From the get go I can notice that the main dataframe that I will be working with is the books data.
Books data devided into 3 columns and connected to ratings and reviews by book_id, publishers and publisher_id and to authors by author_id.

## Tasks

### Find the number of books released after January 1, 2000

In [9]:
query = """
SELECT
    COUNT(title) AS num_books
FROM books
WHERE publication_date >= '2000-01-01';
"""
pd.io.sql.read_sql(query, con = engine)

num_books
0        821

As you can see, I used a filter to find the books that are released after 01/01/2000 and we have counted that we have 819 books.

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Please, pay attention that we need books released **after January 1, 2000**.</div>

### Find the number of user reviews and the average rating for each book.

In [10]:
query = """
SELECT
    books.book_id,
    books.title,
    COUNT(review_id) AS num_review
FROM books
INNER JOIN reviews USING(book_id)
GROUP BY books.book_id
ORDER BY num_review DESC, title;
"""
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  num_review
0        948                            Twilight (Twilight  #1)           7
1        207                                    Eat  Pray  Love           6
2        299  Harry Potter and the Chamber of Secrets (Harry...           6
3        302  Harry Potter and the Prisoner of Azkaban (Harr...           6
4        497                          Outlander (Outlander  #1)           6
..       ...                                                ...         ...
989      872                                The Social Contract           1
990      930  To Green Angel Tower (Memory  Sorrow  and Thor...           1
991      943               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1           1
992      980                              Wicked: The Grimmerie           1
993      984                  Winter Prey (Lucas Davenport  #5)           1

[994 rows x 3 columns]

In [18]:
query = """
SELECT
    books.book_id,
    books.title,
    AVG(rating) AS avg_rate,
    COUNT(DISTINCT(review_id)) AS num_reviews
FROM books
INNER JOIN reviews USING(book_id)
INNER JOIN ratings USING(book_id)
GROUP BY books.book_id
ORDER BY num_reviews DESC, title;
"""
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  avg_rate  \
0        948                            Twilight (Twilight  #1)  3.662500   
1        207                                    Eat  Pray  Love  3.395833   
2        299  Harry Potter and the Chamber of Secrets (Harry...  4.287500   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...  4.414634   
4        497                          Outlander (Outlander  #1)  4.125000   
..       ...                                                ...       ...   
989      872                                The Social Contract  3.500000   
990      930  To Green Angel Tower (Memory  Sorrow  and Thor...  4.500000   
991      943               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1  2.500000   
992      980                              Wicked: The Grimmerie  3.500000   
993      984                  Winter Prey (Lucas Davenport  #5)  4.500000   

     num_reviews  
0              7  
1              6  
2              6  
3              6  
4              6  
..           ...  
989            1  
990            1  
991            1  
992            1  
993            1  

[994 rows x 4 columns]

In [12]:
query = """
SELECT
    books.book_id,
    books.title,
    AVG(rating) AS avg_rate,
    COUNT(DISTINCT(rating_id)) AS num_rate
FROM books
INNER JOIN ratings USING(book_id)
GROUP BY books.book_id
ORDER BY avg_rate DESC, num_rate DESC, title;
"""
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  avg_rate  \
0         17                      A Dirty Job (Grim Reaper  #1)      5.00   
1        553            School's Out—Forever (Maximum Ride  #2)      5.00   
2        347  In the Hand of the Goddess (Song of the Liones...      5.00   
3        444       Moneyball: The Art of Winning an Unfair Game      5.00   
4         20              A Fistful of Charms (The Hollows  #4)      5.00   
..       ...                                                ...       ...   
995      915  The World Is Flat: A Brief History of the Twen...      2.25   
996      202                                      Drowning Ruth      2.00   
997      316                  His Excellency: George Washington      2.00   
998      371                                              Junky      2.00   
999      303                               Harvesting the Heart      1.50   

     num_rate  
0           4  
1           4  
2           3  
3           3  
4           2  
..        ...  
995         4  
996         3  
997         2  
998         2  
999         2  

[1000 rows x 4 columns]

Our task was to find out how many user reviews do we have for each book.

From our table we can see that the most reviewable book is "Twilight" with 7 reviews and after it we have a bunch of books with 6 reviews.

Also, we had a task to find out what is the average rating for each book.

We have found out that the average rating is 5 where the best performarcers with 4 ratings are:
1. A Dirty Job (Grim Reaper #1) 
2. School's Out—Forever (Maximum Ride #2)

### Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [13]:
query = """
SELECT
    books.publisher_id,
    publishers.publisher,
    COunt(books.title) as num_books
FROM publishers
INNER JOIN books on publishers.publisher_id = books.publisher_id
WHERE books.num_pages > 50
GROUP BY books.publisher_id, publishers.publisher
ORDER By num_books DESC;
"""
pd.io.sql.read_sql(query, con = engine)

publisher_id                              publisher  num_books
0             212                          Penguin Books         42
1             309                                Vintage         31
2             116               Grand Central Publishing         25
3             217                       Penguin Classics         24
4              33                       Ballantine Books         19
..            ...                                    ...        ...
329           148          Harvard Business Review Press          1
330           302               Tyndale House Publishers          1
331            66                        Corgi Childrens          1
332           230  Pocket Books / Simon & Schuster  Inc.          1
333           199                          Orchard Books          1

[334 rows x 3 columns]

The publisher that released the greatest number of books with more than 50 pages is "Penguin Books".
He has released 42 books.

### Identify the author with the highest average book rating (look only at books with at least 50 ratings).

In [14]:
query = """
SELECT
    books.author_id,
    authors.author,
    books.title,
    AVG(ratings.rating) AS avg_rate,
    COUNT(ratings.rating_id) AS num_rate
FROM books
    INNER JOIN authors ON books.author_id = authors.author_id
    INNER JOIN ratings ON books.book_id = ratings.book_id
GROUP BY books.author_id,
    books.title,
    authors.author
HAVING COUNT(ratings.rating_id) >= 50
ORDER BY avg_rate DESC;
"""
pd.io.sql.read_sql(query, con = engine)

author_id                                             author  \
0         236                         J.K. Rowling/Mary GrandPré   
1         240                                     J.R.R. Tolkien   
2         236                         J.K. Rowling/Mary GrandPré   
3         402                  Markus Zusak/Cao Xuân Việt Khương   
4         236                         J.K. Rowling/Mary GrandPré   
5         376                                  Louisa May Alcott   
6         236                         J.K. Rowling/Mary GrandPré   
7         240                                     J.R.R. Tolkien   
8          39                                      Arthur Golden   
9         498                                       Rick Riordan   
10        621                                    William Golding   
11        106                                          Dan Brown   
12        235                                      J.D. Salinger   
13        469           Paulo Coelho/Alan R. Clarke/Özdemir İnce   
14        630  William Shakespeare/Paul Werstine/Barbara A. M...   
15        372                                         Lois Lowry   
16        195            George Orwell/Boris Grabnar/Peter Škerl   
17        106                                          Dan Brown   
18        554                                    Stephenie Meyer   
19        311                                     John Steinbeck   

                                                title  avg_rate  num_rate  
0   Harry Potter and the Prisoner of Azkaban (Harr...  4.414634        82  
1   The Fellowship of the Ring (The Lord of the Ri...  4.391892        74  
2   Harry Potter and the Chamber of Secrets (Harry...  4.287500        80  
3                                      The Book Thief  4.264151        53  
4   Harry Potter and the Half-Blood Prince (Harry ...  4.246575        73  
5                                        Little Women  4.192308        52  
6   Harry Potter and the Order of the Phoenix (Har...  4.186667        75  
7                 The Hobbit  or There and Back Again  4.125000        88  
8                                 Memoirs of a Geisha  4.107143        56  
9   The Lightning Thief (Percy Jackson and the Oly...  4.080645        62  
10                                  Lord of the Flies  3.901408        71  
11             The Da Vinci Code (Robert Langdon  #2)  3.830508        59  
12                             The Catcher in the Rye  3.825581        86  
13                                      The Alchemist  3.789474        57  
14                                   Romeo and Juliet  3.787879        66  
15                          The Giver (The Giver  #1)  3.750000        56  
16                                        Animal Farm  3.729730        74  
17               Angels & Demons (Robert Langdon  #1)  3.678571        84  
18                            Twilight (Twilight  #1)  3.662500       160  
19                                    Of Mice and Men  3.622951        61

The author with the highest average book rating with at least 50 pages(or above) is J.K. Rowling/Mary GrandPré with the "Harry Potter and the Prisoner of Azkaban" book(with an average of 4.414 rating and 82 ratings).

### Find the average number of text reviews among users who rated more than 50 books. 

In [15]:
query = """
SELECT
    AVG(COUNT(DISTINCT(reviews.review_id))) OVER () AS avg_reviews_num
FROM reviews
    INNER JOIN ratings ON reviews.username = ratings.username
GROUP By ratings.username
HAVING COUNT(DISTINCT(ratings.rating_id)) > 50
LIMIT 1;
"""
pd.io.sql.read_sql(query, con = engine)

avg_reviews_num
0        24.333333

The average number of text reviews among users who rated more that 50 books is 24.3 rating per user.

## Describe your conclusions for each of the tasks.

- As we gathered our data we could notice from the beginning that the data is formulated good with no issues in it so it was easier to work on it.
- We have found out that there were 819 books released since 01/01/2000.
- We have found out that the average rating is 5 where the best performarcers with 4 ratings are, A Dirty Job (Grim Reaper #1) and School's Out—Forever (Maximum Ride #2).
- Also, we have discovered from our table that the most reviewable book is "Twilight" with 7 reviews and after it we have a bunch of books with 6 reviews.
- The publisher that released the greatest number of books with more than 50 pages is "Penguin Books" with over 40 books(42 to the exact).
- The author with the highest average book rating with at least 50 pages(or above) is J.K. Rowling/Mary GrandPré with the "Harry Potter and the Prisoner of Azkaban" book(with an average of 4.414 rating and 82 ratings).
- The average number of text reviews among users who rated more that 50 books is 24.3 rating per user.